packages/type-storage/src/utils/createFunction.ts

```diff
@@ -0,0 +1,62 @@
+// Copyright 2017-2018 @polkadot/storage authors & contributors
+// This software may be modified and distributed under the terms
+// of the ISC license. See the LICENSE file for details.
+
+import { IStorageFunctionMetadata } from '@polkadot/api-codec/Metadata';
+import String from '@polkadot/api-codec/String';
+import u8aConcat from '@polkadot/util/u8a/concat';
+import xxhash from '@polkadot/util-crypto/xxhash/asU8a';
+
+import { StorageFunction } from '../types';
+
+export interface CreateItemOptions {
+  isUnhashed?: boolean;
+}
+
+/**
+ * From the schema of a function in the module's storage, generate the function
+ * that will return the correct storage key.
+ *
+ * @param schema - The function's definition schema to create the function from.
+ * The schema is taken from state_getMetadata.
+ * @param options - Additional options when creating the function. These options
+ * are not known at runtime (from state_getMetadata), they need to be supplied
+ * by us manually at compile time.
+ */
+export function createFunction (
+  prefix: String,
+  functionMetadata: IStorageFunctionMetadata,
+  options: CreateItemOptions = {}
+) {
+  let func: any; // The return value, it's a function that has metadata.
+
+  if (options.isUnhashed) {
+    func = () => prefix.toU8a();
+  } else {
+    // TODO Find better type than any
+    // Can only have zero or one argument:
+    // - storage.balances.freeBalance(address)
+    // - storage.timestamp.blockPeriod()
+    func = (arg?: any) => {
+      if (functionMetadata.type.isMap) {
+        if (!arg) {
+          throw new Error(`${functionMetadata.name} expects one argument.`);
+        }
+
+        // TODO: Ideal syntax
+        const argType = new functionMetadata.type.asMap.key();
+        return xxhash(u8aConcat(prefix.toU8a(), argType.from(arg).toU8a()), 128);
```

Comment:
Ideally we would want:
- `storage.balances.freeBalance('5ab.....')`
- and `storage.balances.freeBalance([/* equivalent in u8a */])`

to return the same result.

I think a `from` fucntion in CodecBasc that does `if typeof arg === ... then fromU8a else fromJSON` would do the trick.

Reply:
The constructor of `Address` should take care of this case, it takes both u8a & string as params. So `new Address('5C62W7ELLAAfix9LYrcx5smtcffbhvThkM5x7xfMeYXCt72s')` and `new Address([1, 2, 3, 4, 5, 6, 7, 8, 1, 2, 3, 4, 5, 6, 7, 8, 1, 2, 3, 4, 5, 6, 7, 8, 1, 2, 3, 4, 5, 6, 7, 8])` will yields the same.